pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -98,88 +106,243 @@ func (r *MachineSetSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// This function is currently a placeholder.
 	return ctrl.Result{}, nil
 }
 
-// reconcileMAPIMachineSettoCAPIMachineSet MAPI MachineSet to a CAPI MachineSet.
-func (r *MachineSetSyncReconciler) reconcileMAPIMachineSettoCAPIMachineSet(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, capiMachineSet *capiv1beta1.MachineSet) (ctrl.Result, error) {
+// reconcileCAPIMachineSetToMAPIMachineSet reconciles a CAPI MachineSet to a
+// MAPI MachineSet.
+//
+// TODO: Platform specific implementation (currently this works only for AWS,
+// we want a switch on platform somewhere).
+// TODO: Put Gets() for Cluster + Template in helper func.
+func (r *MachineSetSyncReconciler) reconcileCAPIMachineSetToMAPIMachineSet(ctx context.Context, capiMachineSet *capiv1beta1.MachineSet, mapiMachineSet *machinev1beta1.MachineSet) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithValues("function", "reconcileCAPIMachineSetToMAPIMachineSet")
+
+	// Get the AWSCluster
+	cluster := &awscapiv1beta1.AWSCluster{}
+	clusterNamespacedName := client.ObjectKey{
+		Namespace: capiMachineSet.GetNamespace(),
+		Name:      capiMachineSet.Spec.ClusterName,
+	}
+
+	err := r.Get(ctx, clusterNamespacedName, cluster)
+	if err != nil {
+		logger.Error(err, "Failed to get CAPI Cluster", "cluster", clusterNamespacedName.Name)
+
+		r.updateSynchronisedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToGetCluster", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to get CAPI Cluster: %w", err)
+	}
+
+	// Get the AWSMachineTemplate
+	template := &awscapiv1beta1.AWSMachineTemplate{}
+	templateNamespacedName := client.ObjectKey{
+		Namespace: capiMachineSet.Spec.Template.Spec.InfrastructureRef.Namespace,
+		Name:      capiMachineSet.Spec.Template.Spec.InfrastructureRef.Name,
+	}
+
+	err = r.Get(ctx, templateNamespacedName, template)
+	if err != nil {
+		logger.Error(err, "Failed to get AWSMachineTemplate", "name", templateNamespacedName.Name)
+
+		r.updateSynchronisedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToGetAWSMachineTemplate", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to get AWSMachineTemplate: %w", err)
+	}
+
+	// Convert the CAPI MachineSet and AWS resources to a MAPI MachineSet
+	convertedMachineSet, warns, err := capi2mapi.FromMachineSetAndAWSMachineTemplateAndAWSCluster(
+		capiMachineSet, template, cluster,
+	).ToMachineSet()
+
+	if err != nil {
+		logger.Error(err, "Failed to convert CAPI MachineSet to MAPI MachineSet")
+
+		r.updateSynchronisedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToConvertMachineSet", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to convert CAPI MachineSet to MAPI MachineSet: %w", err)
+	}
+
+	for _, warning := range warns {
+		logger.Info("Warning during conversion", "warning", warning)
```

Comment:
I wonder if there's a good way to expose these to end users. I would expect perhaps to use an event recorder to expose these at a minimum

Reply:
I'll expose events alongside the log. 

The only other options I think we have are:

 - Adding an additional condition, but surfacing warnings in a condition feels like we're deviating from the intended use of conditions
 - Adding an additional field to the Machine/Machineset status, but that means additional API changes

Reply:
I'm happy for warnings to stay as events for now until we find a need for them not to be